doc/api/vm.md

```diff
@@ -43,6 +43,319 @@ console.log(x); // 1; y is not defined.
 *Note*: The vm module is not a security mechanism.
 **Do not use it to run untrusted code**.
 
+## Class: vm.Module
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1 - Experimental
+
+The `vm.Module` class provides a low-level interface for using ECMAScript
+modules in VM contexts. It is the counterpart of the `vm.Script` class that
+closely mirrors [Source Text Module Record][]s as defined in the ECMAScript
+specification.
+
+Unlike `vm.Script` however, every `vm.Module` object is bound to a context from
+its creation. Operations on `vm.Module` objects are intrinsically asynchronous,
+in contrast with the synchronous nature of `vm.Script` objects. With the help
+of async functions, however, manipulating `vm.Module` objects are fairly
+straightforward.
+
+Using a `vm.Module` object requires four distinct steps: creation/parsing,
+linking, instantiating, and evaluation. These four steps are illustrated in the
+following example.
+
+This feature is only available with the `--experimental-vm-modules` command
+flag enabled.
+
+*Note*: This implementation lies at a lower level than the [ECMAScript Module
+loader][]. There is also currently no way to interact with the Loader, though
+support is planned.
+
+```js
+const vm = require('vm');
+
+const contextifiedSandbox = vm.createContext({ secret: 42 });
+
+(async () => {
+  // Step 1
+  //
+  // Create a Module by constructing a new `vm.Module` object. This parses the
+  // provided source text, throwing a `SyntaxError` if anything goes wrong. By
+  // default, a Module is created in the top context. But here, we specify
+  // `contextifiedSandbox` as the context this Module belongs to.
+  //
+  // Here, we attempt to obtain the default export from the module "foo", and
+  // put it into local binding "secret".
+
+  const bar = new vm.Module(`
+    import secret from 'foo';
+    secret;
+  `, { context: contextifiedSandbox });
+
```

Comment:
Can the `contextifiedSandbox` for the `vm.Module` assigned to `bar` be different the one used in the `linker` call of `vm.Module` for `foo`?

Reply:
all modules that are linked together **must** be in the same context

Reply:
Maybe doing `import s from 'foo'` then so it's not the same name as the global.

Reply:
> all modules that are linked together **must** be in the same context

Could that be made available to the `linker` as an arg or a `this.context` so it doesn't have to be passed around as an var enclosed in an outer scope.

Reply:
i don't understand what you mean

Reply:
At the moment you have to define context in an outer scope, but since it's required to be the same it might as well be tracked by the created module so things like the `linker` will have access to it without needing to have the outer scope var.

Reply:
> Maybe doing `import s from 'foo'` then so it's not the same name as the global.

:+1:

> At the moment you have to define context in an outer scope, but since it's required to be the same anyways it might as well be tracked by the created module so things like the `linker` will have access to it without needing to have access to the outer scope var.

The linker function already has access to it through `referencingModule.context`.

Reply:
> The linker function already has access to it through `referencingModule.context`.

Nice! Can the example be updated to highlight it.

Reply:
Fixed.